File: solutions/go/need-for-speed/1/need_for_speed.go
package speed

type Car struct {
	battery      int
	batteryDrain int
	speed        int
	distance     int
}

func NewCar(speed, batteryDrain int) Car {
	var c Car
	c.battery = 100
	c.batteryDrain = batteryDrain
	c.speed = speed
	return c
}

type Track struct {
	distance int
}

func NewTrack(distance int) Track {
	var t Track
	t.distance = distance
	return t
}

func Drive(car Car) Car {
	battery := car.battery - car.batteryDrain
	if battery >= 0 {
		car.battery -= car.batteryDrain
		car.distance += car.speed
	}
	return car
}

func CanFinish(car Car, track Track) bool {
	cycles := car.battery / car.batteryDrain
	return car.speed*cycles >= track.distance
}

File: solutions/java/need-for-speed/1/src/main/java/NeedForSpeed.java
class NeedForSpeed {
    private final int BATTERY_FULL = 100;

    private int battery = BATTERY_FULL;
    private int batteryDrain;
    private int speed;
    private int distance = 0;

    public NeedForSpeed(int speed, int batteryDrain) {
        this.speed = speed;
        this.batteryDrain = batteryDrain;
    }

    public boolean batteryDrained() {
        return battery <= 0;
    }

    public int distanceDriven() {
        return distance;
    }

    public void drive() {
        if (!batteryDrained()) {
            battery -= batteryDrain;
            distance += speed;
        }
    }

    public static NeedForSpeed nitro() {
        return new NeedForSpeed(50, 4);
    }
}

class RaceTrack {
    private int length;

    public RaceTrack(int length) {
        this.length = length;
    }

    public boolean tryFinishTrack(NeedForSpeed car) {
        while (car.distanceDriven() < length) {
            if (car.batteryDrained()) {
                return false;
            }
            car.drive();
        }
        return true;
    }
}
